### DISCUSSION TITLE
Page scroll depth

### DISCUSSION THREAD

**Discussion by metmarkosaric** [plausible/analytics#121] [Anything Else]
"Show how far down people scroll on a page as a percentage."

**Update:** Scroll depth tracking is now live https://plausible.io/docs/scroll-depth

**Comment by jkettmann** [comment]
Would be great to get this feature. Especially for blog pages. Any updates?

**Reply by calicojack** [reply]
This is a big one for me, as we use it for a quality measure of the content and engagement.

**Comment by Totalus** [comment]
That would be a very nice feature to add, and such an easy one to implement I believe.

Here's a scrolling depth tracking script as reference:

```js
var scrollDepth = 0;
window.addEventListener("scroll", function (event) {
  var scrollPosition = this.scrollY;
  var maxScrollHeight = document.documentElement.scrollHeight - document.documentElement.clientHeight;

  if(maxScrollHeight == 0) {
    return; // Page too short to scroll
  }

  var scrollPercentage = Math.round(scrollPosition / maxScrollHeight * 100);
  console.log("Scroll position:", scrollPercentage, "%");
  
  if(scrollDepth < scrollPercentage) {
    scrollDepth = scrollPercentage;
    console.log("Scroll depth (max):", scrollDepth);
    
    // TODO : notify Plausible that the user has scrolled deeper
  }
});
```
You could add a debounce mechanism if you want to optimize the number of events sent to the server (ex: send the new depth after the user has sit still for 2 seconds).

On the backend side we simply need to create a new entry in the database for this value and display it in the _Top pages_ detailed table.

What do you think ?

**Comment by foligo** [comment]
Hey there, this is a great request. Are there any updates on that? @metmarkosaric thx a lot in advance.

**Comment by metmarkosaric** [comment]
thanks! there are no short term plans for this. we are a small team with limited development resources so need to focus on most frequently requested features first. currently the focus is on introducing a google analytics stats import. when there's an update on page scroll depth, we'll post it here. thanks for understanding!

**Reply by Totalus** [reply]
How much time would that take to implement ? Seems like an easy to add, yet good value feature.

**Reply by metmarkosaric** [reply]
for every new feature there are many things to think about including how does it affect the size of our script, where to place it in the UI in order to keep having a simple interface, make it look good on all devices, include it in the export and the api etc. we have a small team so need to take things one at a time. at the moment, we're focused on more highly requested features

**Reply by NiklasBr** [reply]
> there are no short term plans for this

@metmarkosaric has there been any change to this over the past years?

**Reply by MickL** [reply]
Two years later, any update on this? :)

**Comment by alexcombessie** [comment]
Hi,

Would love to have "scroll" as a goal event, regardless of scroll depth. It is native in google analytics, I am looking for a similar native feature.

Is it possible today with a bit of custom code? Is it on your roadmap?

Many thanks,

Alex

**Reply by ukutaht** [reply]
Definitely possible with custom code and our Custom events feature. Automatic support for this is not really on the horizon in our roadmap at the moment.

**Reply by Totalus** [reply]
Yes, I'm currently doing that with custom events. It's not that great because each depth count as an individual event, but it works.

Here a code snippet, similar to the one I shared earlier, that I use for that. This one only notifies plausible only each 10% depth to avoid having too much distinct events.

```html
 <!-- Detect scroll down on pages -->
 <script>          
     var scrollDepth = 0;
     window.addEventListener("scroll", function (event) {
         // Called on scroll event
         var scrollPosition = this.scrollY;
         var maxScrollHeight = document.documentElement.scrollHeight - document.documentElement.clientHeight;

         if(maxScrollHeight == 0) {
             // Page too short to scroll
             return;
         }

         var scrollPercentage = Math.round(scrollPosition / maxScrollHeight * 100);
         var scrollStep = 10; // Notify new depth each step
         scrollPercentage = scrollPercentage - (scrollPercentage % scrollStep);
         
         if(scrollDepth < scrollPercentage) {
             // User has scrolled deeper
             scrollDepth = scrollPercentage;

             // Use plausible custom event to capture the scrolling depth
             plausible("Scrolldown", {
                 props: {
                     path:window.location.pathname,
                     depth: scrollDepth,
                     tag: window.location.pathname + " " + scrollDepth.toString() + "%"
                 },
             })
         }
     });
 </script>
```

This gives you a _Scrolldown_ goal that displays an event for each 10% scrolldown of each page.

![image](https://user-images.githubusercontent.com/34516402/154528725-348cdafd-2e1d-4fb3-b3e9-3e7ff45ecd9c.png)

Here you can see the root page of my website was scrolled to 100% by one person. You'll get different counts depending on how many users scroll different depths.

See : [Custom event goals](https://plausible.io/docs/custom-event-goals) for more info.

**Reply by alexcombessie** [reply]
You just made my day, thank you so much @Totalus !

**Reply by metmarkosaric** [reply]
thanks for sharing @Totalus!

**Reply by dgp1130** [reply]
Is there any way to limit these results to the maximum scroll for a single user / session? A user who scrolls 50% also scrolls 10%, 20%, etc. which introduces a lot of noise. I guess these numbers represent "X users scrolled *at least* Y% of the page". AFAICT there's no way to generate "X users scrolled between Y% and Y+10% of the page". This makes it harder to identify where users are falling off.

I guess you could subtract adjacent buckets and say "If 15 users scrolled 30% and 10 users scrolled 40%, then 5 users must have scrolled between 30% - 39%". It would nice to express that in the visualization though.

**Reply by Totalus** [reply]
You probably can if you only notify plausible when the user exits the page instead of when the depth changes, something like :

```js
window.addEventListener('unload', notifyPlausible, false);

function notifyPlausible() {
     plausible("Scrolldown", {
         props: {
             path:window.location.pathname,
             depth: scrollDepth,
             tag: window.location.pathname + " " + scrollDepth.toString() + "%"
         },
     })
}
```

This might not work 100% of the time however. You'd have to test, but I think if the user exits the page by closing the tab or exiting the browser, the request won't be made.

**Reply by abett** [reply]
I think that would be too much of a constraint for me to rely on it..
But it'd be a perfect application for [navigator.sendBeacon](https://developer.mozilla.org/en-US/docs/Web/API/Navigator/sendBeacon), wouldn't it?
How's the roadmap for https://github.com/plausible/plausible-tracker/issues/16?

**Reply by metmarkosaric** [reply]
you can see the current status in this PR https://github.com/plausible/analytics/pull/1679

**Reply by jkettmann** [reply]
I wrote a similar script that attaches an intersection observer to all h2 and h3 headings and sends an event once a heading becomes visible. Here's my React code in case anyone is interested:

```
useEffect(() => {
    const callback = (headings, observer) => {
      const activeHeading = headings.find((h) => h.isIntersecting);
      if (activeHeading) {
        observer.unobserve(activeHeading.target);
        plausible("Read progress", {
          props: {
            section: `${activeHeading.target.tagName} ${activeHeading.target.textContent}`,
          },
        });
      }
    };

    const observer = new IntersectionObserver(callback, {
      rootMargin: "0px 0px -40% 0px",
    });

    const headingElements = Array.from(document.querySelectorAll("h2, h3"));
    headingElements.forEach((element) => observer.observe(element));

    return () => observer.disconnect();
  });
```

~~The resulting goal could be really valuable, but unfortunately it's not easy to read because of the wrong order of headings.~~

Edit: With a bit of patience and more data the order of headings becomes clearer. It's now quite easy to see how far the users read.

<img width="1045" alt="grafik" src="https://user-images.githubusercontent.com/9722792/154801079-cfae9d1d-7343-4da6-8fb9-869ddc4e5aef.png">

**Reply by dgp1130** [reply]
I implemented mine with a similar approach to https://github.com/plausible/analytics/discussions/121#discussioncomment-2198451 (https://github.com/dgp1130/blog/commit/4659e35baed0bd6caf12a942cba490314f37fbd5 and https://github.com/dgp1130/blog/commit/53f782686a3446897e05bbd4384e9f156c4b2f60 if you're interested). This works well enough for a small blog, but I'm not totally satisfied with the approach since Plausible's dashboard doesn't give a means of breaking down this data. I'd want to filter for a specific page and view how far users scrolled down it, but there is no way to do that in the custom events dashboard. Instead we have to make a `tag` property which contains both the path and the scroll depth of the event, which intermixes pages and can get quite unwieldy. Ideally I would be able to:

1. Filter to `scroll-depth.path == "/some/page"`.
2. View the `scroll-depth.depth` property.
3. Sort the result ascending.

The simplest workaround would be to view the `tag` property and then just add a filter which limits to `tag` values that contain `/some/page`, but I don't think even that's possible right now. That would probably be the easiest way to support this feature, though some more powerful dashboard views would do a more complete job.

**Comment by johannbuscail** [comment]
Any update on this ?

**Comment by nelsonic** [comment]
We are suuper keen on this feature too. Happy to _contribute_ code for it. 🙏 
We are current paying customers of Plausible and planning to propose using the product to a large corporate client that needs privacy. 
But this is a feature they they need and thus we are either going to `fork` the `analytics` project and add it to the `fork` 
***or*** we'd be very happy to contribute it up-stream and have it in the core product.

**Comment by metmarkosaric** [comment]
**Update:** Scroll depth tracking is now live https://plausible.io/docs/scroll-depth

**Reply by NiklasBr** [reply]
Can we migrate a custom property to this?

**Reply by metmarkosaric** [reply]
@NiklasBr as in a custom property you've set up to manually track scroll depth? unfortunately it's not possible to migrate/merge that with the scroll depth tracking feature itself

**Reply by MickL** [reply]
@metmarkosaric Amazing, I am so happy for this feature! However, how should it be implemented? The JS script [plausible-tracker](https://github.com/plausible/plausible-tracker) has not been updated for a long time. Most people are using [@barbapapazes/plausible-tracker](https://github.com/barbapapazes/plausible-tracker) instead I guess. There the page scroll needs to be implemented, but where to find the API docs to do so?

**Reply by NiklasBr** [reply]
> As in a custom property you've set up to manually track scroll depth?

Yes exactly.

**Reply by metmarkosaric** [reply]
@MickL thanks, glad to hear that! scroll depth tracking works out of the box with our default JavaScript snippet and is tracked with a new kind of special event which requires a pretty complex setup. we plan to work on a new version of the tracker to support this but i cannot say any details nor promise any timelines on this. sorry about that!

**Reply by MickL** [reply]
Ok, unfortunately most projects, including [nuxt-plausible](https://github.com/nuxt-modules/plausible) use [barbapapazes/plausible-tracker](https://github.com/Barbapapazes/plausible-tracker) which is an own implementation. To support scrolldepth tracking it would require barbapapazes/plausible-tracker to rewrite this "complex setup". Issue: https://github.com/Barbapapazes/plausible-tracker/issues/20

Would you be willing to show how to make this setup and document the needed api event?

Btw if you ever want to start workinh on plausible-tracker againg, maybe it would make sense to fork Barbapapazes/plausible-tracker instead and use it as a new base?

**Reply by Swamp2k9** [reply]
I've implemented the WordPress plug-in on my sites, but I'm unable to see the scroll percentage values on my Plausible dashboard. 

However, I can view these values on the Plausible test sites. Could someone help me resolve this issue?

**Reply by metmarkosaric** [reply]
hi @Swamp2k9! please do contact our customer support with details and someone will take a look. see https://plausible.io/contact. thanks!

**Reply by MickL** [reply]
@metmarkosaric Could you please provide information on how to do the scroll tracking, maybe in this issue https://github.com/Barbapapazes/plausible-tracker/issues/20 ?

**Comment by ksully-execute** [comment]
I know we don't track users (Privacy-first), and am wondering if this feature could be used for rewarding authors (depending on scroll length + time on page = payment to authors....is this something of interest to anyone or is this unacceptable use?
For my situation, I am only talking about authenticated users, so a cookie would be there from authentication with an opaque one-time value that could be used for correlation....?
